lib/internal/modules/cjs/loader.js

```diff
@@ -297,8 +327,58 @@ function findLongestRegisteredExtension(filename) {
   return '.js';
 }
 
+// This only applies to requests of a specific form:
+// 1. name/.*
+// 2. @scope/name/.*
+const EXPORTS_PATTERN = /^((?:@[^./@\\][^/@\\]*\/)?[^@./\\][^/\\]*)(\/.*)$/;
```

Comment:
does this only go 1 level deep, it seems like `x/y/z/foo` would not check for `y` or `z` having a `package.json` ? Also, why is `@/` not allowed?

Reply:
also seems like expansion should support `\\` as a leading char given the char class before it.

Reply:
`@/` may be possible but I would rather keep it tightly constraint at the beginning and loosen in the future.

> also seems like expansion should support `\\` as a leading char given the char class before it.

I tried double-checking each group and I *think* none supports `\\`. Maybe I'm blind. ^^

Reply:
> does this only go 1 level deep, it seems like `x/y/z/foo` would not check for `y` or `z` having a `package.json`?

That is correct. The reasoning is that this isn't a directory operation but a name operation and we currently assume that names only contain at most one slash and only if the thing before the slash starts with an `@`. We could start checking other package names ("name": "x/y") but then `x` could try to hide internals and accidentally leak them because they added a `package.json` in `y/` for `"type": "module"`. I like the property of "every bare specifier has at most *one* non-ambiguous exports map".

Reply:
Right now it looks like you are correct, but does that mean it doesn't work with currently supported `\\` separation on win32?

Reply:
@jkrems depth limit seems fine but could we put that in docs somewhere, just telling people that they need to setup exports for any nested packages within theirs. Essentially, telling them to whitelist even nested exports. This falls under their package being responsible for its contents so depth is probably fine.

Reply:
Do you mean a sentence like the following following (or within) the require algorithm:

> This implies that `exports` is only honored when loading a package "name" as defined above. Any "exports" values within nested directories are ignored.

Reply:
A sentence, avoid the word "implies". I'd probably trim it down a bit:

> exports is only honored when loading a package "name" as defined above. Any "exports" values within nested directories and packages must be declared by "name".

Reply:
Updated in https://github.com/nodejs/node/pull/28759/commits/3316fea299ec21778afdab79ab73bf3a6ec83d80. I found 'declared by "name"' a bit hard to interpret so I threw in a mention of the `package.json` file.